File: src/helper/EmbedBuilder.ts
import {
  EmbedBuilder as DiscordEmbedBuilder,
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  type ColorResolvable,
  type APIEmbedField,
  type MessageActionRowComponentBuilder,
} from "discord.js";

interface ButtonConfig {
  label: string;
  customId: string;
  style?: ButtonStyle;
  url?: string;
  disabled?: boolean;
}

export class EmbedBuilder {
  private static readonly colors = {
    primary: "#5865F2" as ColorResolvable,
    success: "#57F287" as ColorResolvable,
    warning: "#FEE75C" as ColorResolvable,
    error: "#ED4245" as ColorResolvable,
    info: "#00A8FC" as ColorResolvable,
  };

  private embed: DiscordEmbedBuilder;
  private buttons: ButtonConfig[] = [];

  constructor() {
    this.embed = new DiscordEmbedBuilder();
  }

  static success(title?: string, description?: string): EmbedBuilder {
    return new EmbedBuilder().setSuccess(title, description);
  }

  static error(title?: string, description?: string): EmbedBuilder {
    return new EmbedBuilder().setError(title, description);
  }

  static info(title?: string, description?: string): EmbedBuilder {
    return new EmbedBuilder().setInfo(title, description);
  }

  static warning(title?: string, description?: string): EmbedBuilder {
    return new EmbedBuilder().setWarning(title, description);
  }

  static primary(title?: string, description?: string): EmbedBuilder {
    return new EmbedBuilder().setPrimary(title, description);
  }

  setPrimary(title?: string, description?: string): this {
    this.embed
      .setColor(EmbedBuilder.colors.primary)
      .setTitle(title ?? null)
      .setDescription(description ?? null);
    return this;
  }

  setSuccess(title?: string, description?: string): this {
    this.embed
      .setColor(EmbedBuilder.colors.success)
      .setTitle(title ?? null)
      .setDescription(description ?? null);
    return this;
  }

  setWarning(title?: string, description?: string): this {
    this.embed
      .setColor(EmbedBuilder.colors.warning)
      .setTitle(title ?? null)
      .setDescription(description ?? null);
    return this;
  }

  setError(title?: string, description?: string): this {
    this.embed
      .setColor(EmbedBuilder.colors.error)
      .setTitle(title ?? null)
      .setDescription(description ?? null);
    return this;
  }

  setInfo(title?: string, description?: string): this {
    this.embed
      .setColor(EmbedBuilder.colors.info)
      .setTitle(title ?? null)
      .setDescription(description ?? null);
    return this;
  }

  // ── Author ───────────────────────────────────────────

  setAuthor(name: string, url?: string, iconURL?: string): this {
    this.embed.setAuthor({ name, url, iconURL });
    return this;
  }

  // ── Body ─────────────────────────────────────────────

  setBody(title: string, description: string, url?: string): this {
    this.embed.setTitle(title).setDescription(description);
    if (url) this.embed.setURL(url);
    return this;
  }

  // ── Images ───────────────────────────────────────────

  setImageUrl(url: string): this {
    this.embed.setImage(url);
    return this;
  }

  setThumbnailUrl(url: string): this {
    this.embed.setThumbnail(url);
    return this;
  }

  // ── Footer ───────────────────────────────────────────

  setFooter(text: string, iconURL?: string, timestamp?: Date | boolean): this {
    this.embed.setFooter({ text, iconURL });
    if (timestamp) {
      this.embed.setTimestamp(timestamp === true ? new Date() : timestamp);
    }
    return this;
  }

  // ── Fields ───────────────────────────────────────────

  addInlineField(name: string, value: string, inline?: boolean): this {
    this.embed.addFields({ name, value, inline: inline ?? true });
    return this;
  }

  addBlankField(inline?: boolean): this {
    this.embed.addFields({ name: "\u200B", value: "\u200B", inline: inline ?? false });
    return this;
  }

  // ── Buttons ──────────────────────────────────────────

  addButton(label: string, customId: string, style?: ButtonStyle): this {
    this.buttons.push({ label, customId, style });
    return this;
  }

  addLinkButton(label: string, url: string): this {
    this.buttons.push({ label, url, customId: "", style: ButtonStyle.Link });
    return this;
  }

  // ── Color ────────────────────────────────────────────

  setColor(color: ColorResolvable): this {
    this.embed.setColor(color);
    return this;
  }

  // ── Build ────────────────────────────────────────────

  getEmbed(): DiscordEmbedBuilder {
    return DiscordEmbedBuilder.from(this.embed);
  }

  toMessageOptions() {
    const components: ActionRowBuilder<MessageActionRowComponentBuilder>[] = [];

    if (this.buttons.length > 0) {
      const row = new ActionRowBuilder<MessageActionRowComponentBuilder>();

      for (const btn of this.buttons) {
        const button = new ButtonBuilder()
          .setLabel(btn.label)
          .setStyle(btn.style ?? ButtonStyle.Primary);

        if (btn.url) {
          button.setURL(btn.url);
        } else {
          button.setCustomId(btn.customId);
        }

        if (btn.disabled) button.setDisabled(true);
        row.addComponents(button);
      }

      components.push(row);
    }

    return {
      embeds: [this.getEmbed()],
      components,
    };
  }
}

File: utils/embed_builder.py
import discord
from datetime import datetime
from typing import Optional


class EmbedBuilder(discord.Embed):
    colors = {
        "primary": 0x5865F2,
        "success": 0x57F287,
        "warning": 0xFEE75C,
        "error": 0xED4245,
        "info": 0x00A8FC,
    }

    @classmethod
    def primary(cls, title: Optional[str] = None, description: Optional[str] = None):
        return cls(color=cls.colors["primary"], title=title, description=description)

    @classmethod
    def success(cls, title: Optional[str] = None, description: Optional[str] = None):
        return cls(color=cls.colors["success"], title=title, description=description)

    @classmethod
    def warning(cls, title: Optional[str] = None, description: Optional[str] = None):
        return cls(color=cls.colors["warning"], title=title, description=description)

    @classmethod
    def error(cls, title: Optional[str] = None, description: Optional[str] = None):
        return cls(color=cls.colors["error"], title=title, description=description)

    @classmethod
    def info(cls, title: Optional[str] = None, description: Optional[str] = None):
        return cls(color=cls.colors["info"], title=title, description=description)

    @classmethod
    def hex(cls, hex_color: str, title: Optional[str] = None, description: Optional[str] = None):
        hex_color = hex_color.lstrip("#")
        return cls(color=int(hex_color, 16), title=title, description=description)

    def set_primary(self, title: Optional[str] = None, description: Optional[str] = None):
        self.color = self.colors["primary"]
        if title:
            self.title = title
        if description:
            self.description = description
        return self

    def set_success(self, title: Optional[str] = None, description: Optional[str] = None):
        self.color = self.colors["success"]
        if title:
            self.title = title
        if description:
            self.description = description
        return self

    def set_warning(self, title: Optional[str] = None, description: Optional[str] = None):
        self.color = self.colors["warning"]
        if title:
            self.title = title
        if description:
            self.description = description
        return self

    def set_error(self, title: Optional[str] = None, description: Optional[str] = None):
        self.color = self.colors["error"]
        if title:
            self.title = title
        if description:
            self.description = description
        return self

    def set_info(self, title: Optional[str] = None, description: Optional[str] = None):
        self.color = self.colors["info"]
        if title:
            self.title = title
        if description:
            self.description = description
        return self

    def set_color_hex(self, hex_color: str, title: Optional[str] = None, description: Optional[str] = None):
        hex_color = hex_color.lstrip("#")
        self.color = int(hex_color, 16)
        if title:
            self.title = title
        if description:
            self.description = description
        return self

    def set_author(self, name: str, url: Optional[str] = None, icon_url: Optional[str] = None):
        super().set_author(name=name, url=url, icon_url=icon_url)
        return self

    def set_body(self, title: str, description: str, url: Optional[str] = None):
        self.title = title
        self.description = description
        if url:
            self.url = url
        return self

    def set_image_url(self, url: str):
        self.set_image(url=url)
        return self

    def set_thumbnail_url(self, url: str):
        self.set_thumbnail(url=url)
        return self

    def set_footer(self, text: str, icon_url: Optional[str] = None, timestamp=None):
        super().set_footer(text=text, icon_url=icon_url)
        if timestamp:
            self.timestamp = timestamp if isinstance(timestamp, datetime) else discord.utils.utcnow()
        return self

    def add_inline_field(self, name: str, value: str, inline: bool = True):
        self.add_field(name=name, value=value, inline=inline)
        return self

    def add_blank_field(self, inline: bool = False):
        self.add_field(name="\u200B", value="\u200B", inline=inline)
        return self
